### PR TITLE
Replace `Transformation` and `Validation` with new groups of the refinery

### PR DESCRIPTION
**Pull request by legionth**
This PR will get move the transformations and constraints of the `Transformation/Factory` and `Validation/Factory` factory to equivalent groups for the refinery.

**One big pull request**

I was in between the decision to create several small PRs and a big one. I decided to go with a big PR, because there several constraints/transformations that interact with each other.

_If you struggle to review this PR. I could try to separate it to smaller PRs, but I would like to give it a shot first._

**What is new**

* New groups for the refinery were created
  * Container
     * contains the old `AddLabelConstraint`
     * Normally I would have named this group `Array` but this is a reserved name in PHP
  * Custom
    * contains the old `Constraints/Custom` and `Transformation/Constraint`
 * Logical
   * contains all the old constraints that are based on a logical operations (`not`, `or` etc.)
   * The constraint for the logical operation `Or` is named `LogicalOr`, because as the word `Array` this word is reservered in PHP
  * Null
     * contains the old `IsNull` constraint
  * Numeric
     * contains the old `IsNumeric` constraint
  * Password
     * contains all the old constraints for passwords (`HasMinLength`, `HasSpecialChars` etc.)
* Code that already used constraints has been adapted.
   * Please have a look especially on `NewMethodTransformation`. I added this line because we using the `NewMethodTransformation` via `Refinery::to()::data()` instead of `Transformation::data()`. Because we are using the `NewMethodTransformation`, the data that is needed has to be passed as in an array of parameters (caused by the internal usage of `call_user_func_array`). Otherwise tests like `PasswordInputTest::test_value_type` and `PasswordInputTest::test_value_required` would fail because the internal usage of `Password::withInput` will transport these values to `Refinery::to()::data()` as a single value and not wrapped in an array. I added this wrapping step in the transformation method to simulate the old behaviour. I hope this is ok in the first step, otherwise we would need to adapt the `InputData` behavior.
* The READMEs were just merged in the current README. As described in #1707 a follow up PR with a proper README will follow.

### CODE REVIEW

File: src/Refinery/Factory.php [hunk 1]
```diff
@@ -79,4 +77,64 @@ public function string(): String\Group
...
+	}
+
+	/**
+	 * Contains constraints for null types
+	 *
+	 * @return Null\Group
+	 */
+	public function null()
+	{
+		return new Null\Group($this->dataFactory, $this->language);
```
Comment (mjansenDatabay) [inline src/Refinery/Factory.php]: What about a `state` group instead @legionth / @klees ?

Comment (legionth) [reply src/Refinery/Factory.php]: I agree with you `null` is not a perfect name for this, but I'm not sure what `state` actual is. Often the term `state` is also used for boolean states, which can lead to confusion IMO. I have no actual  counter proposal. So further input is much appreciated :+1:

Comment (mjansenDatabay) [reply src/Refinery/Factory.php]: You are right, this could lead to confusion. So IMO `null` is okay for now, maybe we find a better wording in future.

**Review by mjansenDatabay — APPROVED** [review]
LGTM (except my objection regading the `null` group name, but as described I have no better wording but `state`).

Let's continue with a final `README` PR after @klees reviewed this.

**Comment by klees** [comment]
Hi @legionth!

Really nice to see the `Refinery` coming into shape! This review mostly contains suggestions that will polish the stuff further. Feel free to decline them if you are running out of the volume of your initial mandate.

Please answer the following questions:
- [x] Do we still need `Logical\Constraint\Parallel` and `Serial`? I think these are the same then `In\Series` and `In\Parallel`.

Please consider the following suggestions. You do not need to follow those, but but please indicate shortly why you prefer to do otherwise:
- [x] Do we expect any more custom constraints or transformations to arise? If not I would propose to just use `Custom\Constraint` as name instead of `Custom\Constraint\Custom`. Same for `Custom\Transformations\Custom`, lets just use `Custom\Transformation`.
- [x] Let's use `array` instead of `container` as name for the group. What other types of containers would we expect to be transformed by transformations in this group?
- [x] Move `isNull` to the top-level (until we find other things to group it with). A `null` group does not make sense imo. Since `null` is such an uninteresting type/value we won't be able to do much with it...
- [x] Remove that intermediate `Constraint`/`Transformation`-Namespaces e.g. in `String\Transformation\SplitSprint`. From my POV the distinction between transformations and constraints became quite uninteresting since we unified them in one concept with the introduction of the refinery. It should be possible to tell by the name which is which.
- [x] Create a ROADMAP.md (maybe similar to the one of the UI-framework) and put "replace typehints and return types in doc-strings by real typehints and return-types" on it.

Btw.: [this](https://github.com/ILIAS-eLearning/ILIAS/pull/1834/files#diff-4fbc4fe42e200b145f673159c7fcd07cR103) looks amazingly readable.

@Amstutz: I think we need to do some cleanup in the inputs of the UI-framework once this is merged. [This](https://github.com/ILIAS-eLearning/ILIAS/pull/1834/files#diff-4fbc4fe42e200b145f673159c7fcd07cR103) looks as if there is one to much layer and we could just use that constraint instead of evaluating it into another custom constraint. We also may want to unify `withAdditionalConstraint` and `withAdditionalTransformation`.

Best Regards!

**Review by klees — COMMENTED** [review]
See above.

**Comment by legionth** [comment]
Thank you for your feedback :+1: 

> Do we still need Logical\Constraint\Parallel and Serial? I think these are the same then In\Series and In\Parallel.

Basically I agree with you, but the current architecture of `src/UI` doesn't  allow to anything besides `Contraints`.
I agree with you that the definitions of `Transformation` and `Constraints` are blurry and should be some kind of unified. On the current state `Constraint\Parallel` and `Constraint\Serial` and the `In` group equivalents seem identical, but actually aren't at the current state.
These constrains were kept, because they are used in productive code in `src/UI` first.


> Do we expect any more custom constraints or transformations to arise? If not I would propose to just use Custom\Constraint as name instead of Custom\Constraint\Custom. Same for Custom\Transformations\Custom, lets just use Custom\Transformation.

Sure can do. I just wanted to create a common folder structure so everybody can find the desired file fast.

```
+ GroupFolder/
|-+ Group.php
|-+ Transformation/
|-+ Constraints/
```
I'll move the `custom` transformation and constraints to the top level of the group folder.

>  Let's use array instead of container as name for the group. What other types of containers would we expect to be transformed by transformations in this group?

The problem here is that `array` is a special keyword in PHP. While it can be used for folders it can't be used in e.g. functions or namespaces. That could be a problem if the ILIAS project moves to a PSR-4 namespace standard, which basically means the namespace is (kinda) the folder structure.
I used the name `container` because it made the most sense for an alternative here and get rid of the namespace. @mjansenDatabay also proposed `Collection` as an alternative. But would love to get some feedback on this matter :+1: 

>  Remove that intermediate Constraint/Transformation-Namespaces e.g. in String\Transformation\SplitSprint. From my POV the distinction between transformations and constraints became quite uninteresting since we unified them in one concept with the introduction of the refinery. It should be possible to tell by the name which is which.

As mentioned before, my original intention was to create a common folder structure. I see still a difference between `Transformation` and `Constraints` because these are still very different interfaces.

But I will put these to the current top level of each group and adapt the namespace accordingly.

> Create a ROADMAP.md (maybe similar to the one of the UI-framework) and put "replace typehints and return types in doc-strings by real typehints and return-types" on it.

Will do :+1:

**Comment by legionth** [comment]
Latest commit moved `IsNull` to top level of the Refinery as requested.

@mjansenDatabay and I decided to keep the name `Container` over the name `Array_`. I hope this is ok for you @klees . If you want, I can add some explanation, but I wanted to keep it short.

**Comment by klees** [comment]
>  I hope this is ok for you @klees .

Yes, absolutely fine.

**Pull request merged**